compiler/passes/scopeResolve.cpp

```diff
@@ -1474,7 +1475,17 @@ static void setupShadowVars() {
         checkRefsToIdxVars(fs, def, svar);
     }
 
-  // Instead of the two nested loops above, we could march through
+  forv_Vec(BlockStmt, bs, gBlockStmts)
+    if(ForLoop *fl = toForLoop(bs)) {
+      for_shadow_vars_and_defs(svar, def, temp, fl) {
+        if (hasOuterVariable(svar))
+          setupOuterVar(fl, svar);
+        if (svar->isTaskPrivate())
+          checkRefsToIdxVars(fl, def, svar);
+      }
```

Comment:
This is just the second repetition between `forall` and `for`, but the core of the implementation is complex enough to deserve a helper.

Reply:
I assume you're suggesting that I factor out the common body of these two loops and call to that.

Sounds good, I refactored it to this:

```C++
  template<class T>
  static void setupShadowVarsOnLoop(T *loop) {
    INT_ASSERT(isForallStmt(loop) || isForLoop(loop));
    for_shadow_vars_and_defs(svar, def, temp, loop) {
      if (hasOuterVariable(svar))
        setupOuterVar(loop, svar);
       if (svar->isTaskPrivate())
        checkRefsToIdxVars(loop, def, svar);
    }
  }

  static void setupShadowVars() {
    forv_Vec(ForallStmt, fs, gForallStmts)
      setupShadowVarsOnLoop(fs);

    forv_Vec(BlockStmt, bs, gBlockStmts)
      if(ForLoop *fl = toForLoop(bs))
        setupShadowVarsOnLoop(fl);

    // Instead of the nested loops above, we could march through
    // gShadowVarSymbols and invoke setupOuterVar(svar->parentExpr, svar).
    // The nested loops group together all shadow variables of a given
    // ForallStmt, so hopefully have better cache behavior.

    USR_STOP();
  }
````

I guess ideally I could have `setupShadowVarsOnLoop` take a pointer to my new interface and use that but that doesn't seem to play well with `for_shadow_vars_and_defs` which is assuming a specifically named data member (which maybe I should change how that iterator works but I'll wait to see if this comes up a lot more).